### ISSUE TITLE
remote store builds (`--store`) can fail if called against a derivation file

### ISSUE THREAD

**Issue by nrdxp** [NixOS/nix#7220]
**Describe the bug**

I want eval a bunch of CI jobs with `nix-eval-jobs` for faster evaluation, which I then want to run via a remote store using `--store`. Since this command outputs a bunch of already evaluated derivations, there is no eval step required during the build command. However, this seems to cause the build to fail if the `drv` file doesn't already exist on the remote store.

**Steps To Reproduce**

evaluate a nix expression to a derivation:
`nix-instantiate .`
Then call nix build via a remote store:
`nix build --eval-store auto --store ssh-ng://my-store /nix/store/*-derivation-just-produced.drv`

If the derivation has never existed on the target machine, the build will fail.

**Expected behavior**

If using `--eval-store auto --store ssh-ng://my-store` on a nix expression, the derivations that expression produces are uploaded first to the build machine, and then the build is run, ensuring they exist. I see no reason why the same thing can't happen if a nix build is initiated on a derivation file directly.

**`nix-env --version` output**
nix-env (Nix) 2.11.2

**Comment by thufschmitt** [comment]
Agreed, this looks like a bug. Strange though because the logic that copies the derivations to the build store shouldn't depend on the shape of the installable that we want to build

**Comment by thufschmitt** [comment]
Just to confirm: I can reproduce this with

```console
$ nix path-info --derivation nixpkgs#hello
/nix/store/c7jrxqjhdda93lhbkanqfs07x2bzazbm-hello-2.12.drv
$ l /nix/store/c7jrxqjhdda93lhbkanqfs07x2bzazbm-hello-2.12.drv
-r--r--r-- 1 root root 1,4K  1 janv.  1970 /nix/store/c7jrxqjhdda93lhbkanqfs07x2bzazbm-hello-2.12.drv
$ nix build /nix/store/c7jrxqjhdda93lhbkanqfs07x2bzazbm-hello-2.12.drv --store /tmp/store --eval-store auto
error: path '/nix/store/c7jrxqjhdda93lhbkanqfs07x2bzazbm-hello-2.12.drv' is not a valid store path
$ nix build nixpkgs\#hello --store /tmp/store --eval-store auto # Just works
$ 
```

**Comment by nrdxp** [comment]
added a comment to the description but it seems `nix-build` does not have this problem as far as I can tell in my tests so far. A little surprised that the two commands don't use similar code paths.

**Comment by thufschmitt** [comment]
I've a hunch that the culprit is [here](https://github.com/NixOS/nix/blob/9bff7e8ee26c97441858e1cea097f44c6db61235/src/libcmd/installables.cc#L408): This code path is only used by `nix build`, and seems to try to read the derivation from the build store to get its output names (and that happens before we copy the drv files to it).

@Ericson2314 any idea how we could avoid that?

**Comment by Ericson2314** [comment]
The fact that installables have their own `rec<Store>` is weird. `evalStore` was also hacked in without regard to layering (the very name "eval store" is a layer violation in `libstore`: it should be called `drvStore` instead or similar.

I don't see any obvious solutions at a glance.

My hope is that via splitting Nix before stabilizing things per my RFCs gives us some breathing room to work through this stuff and make sure the designs make sense: currently a lot of stuff like this is sketchy and needs more eyeballs / more pondering.

**Comment by Ericson2314** [comment]
@nrdxp 

> I can't use the new cli atm because of https://github.com/NixOS/nix/issues/7220

Did you try making both the eval store and the regular store the remote store?

**Comment by Ericson2314** [comment]
I am working on the `libcmd` code right now trying to implement https://github.com/NixOS/rfcs/pull/134, and yeah this stuff is just a mess. The OOP things are repeatedly causing me trouble, and I am told by my colleagues with more C++ than me that no good professional C++ teams write things this way (using the OOP parts) anymore.

If we want to root out these types of bugs, we should consider gutting this code in a more major way.

**Comment by Ericson2314** [comment]
I will attempt a more narrow fix of this sort of thing make commands but *not* installables down stores, we will see how it goes.

**Comment by Ericson2314** [comment]
There is also going in the opposite direction, which is keeping the installables owning stores/eval states, and getting rid of the stores from the `build` (and similar) methods.

**Comment by nrdxp** [comment]
> If we want to root out these types of bugs, we should consider gutting this code in a more major way.

I am glad I am not the only person who feels this way. My recent adventures in trying to get more familiar with the Nix code have landed me in a very similar boat. Not sure exactly what we should do about it, but I already have some experimental branches going to try my hand at refactoring a few things. Not sure if that will pan out in the end, but if nothing else it will help me to continue to build experience in the codebase up to a point that I can actively contribute to such an effort.

**Comment by blaggacao** [comment]
I just re-reproduced according to @thufschmitt instructions on `2.15`, maybe there is a signal in that the output is now different:
```console
❯ ./result/bin/nix build --eval-store auto --store /tmp/store /nix/store/7ky0zmis8b384k5sx852i0fq7x9ir2jl-hello-2.12.1.drv
error: path '/nix/store/7ky0zmis8b384k5sx852i0fq7x9ir2jl-hello-2.12.1.drv' is required, but there is no substituter that can build it
```

Same for `--eval-store /tmp/store`.
